File: 15.三数之和.cpp
/*
 * @lc app=leetcode.cn id=15 lang=cpp
 *
 * [15] 三数之和
 */
#include "algorithm"
#include "iostream"
#include "map"
#include "queue"
#include "vector"

using namespace std;
// @lc code=start
class Solution {
  public:
    vector<vector<int>> threeSum(vector<int> &nums) {
        std::sort(nums.begin(), nums.end());
        vector<vector<int>> result;
        int max = 0, min = 0, n = nums.size();
        for (auto &i : nums) {
            if (i > max)
                max = i;
            if (i < min)
                min = i;
        }
        map<string, bool> m;
        int sum2;
        for (int i = 0; i < n - 2; i++) {
            for (int j = i + 1; j < n - 1; j++) {
                sum2 = nums[j] + nums[i];
                if ((sum2 > 0 && sum2 + min > 0) ||
                    (sum2 < 0 && sum2 + max < 0)) {
                    continue;
                }
                if (m[to_string(nums[i]) + to_string(nums[j])] &&
                    m[to_string(nums[j]) + to_string(nums[i])]) {
                    continue;
                }
                for (int k = j + 1; k < n; k++) {
                    if (sum2 + nums[k] == 0) {
                        m[to_string(nums[i]) + to_string(nums[j])] = true;
                        m[to_string(nums[i]) + to_string(nums[k])] = true;
                        m[to_string(nums[j]) + to_string(nums[k])] = true;

                        m[to_string(nums[j]) + to_string(nums[i])] = true;
                        m[to_string(nums[k]) + to_string(nums[i])] = true;
                        m[to_string(nums[k]) + to_string(nums[j])] = true;
                        result.push_back({nums[i], nums[j], nums[k]});
                        break;
                    }
                }
            }
        }
        return result;
    }
};
// @lc code=end

File: 15.三数之和.go
/*
 * @lc app=leetcode.cn id=15 lang=golang
 *
 * [15] 三数之和
 */

// @lc code=start


func threeSum(nums []int) [][]int {
	sort.Ints(nums)
	var ret [][]int
	size := len(nums)
	k := 0
	for ; k < size-2; k++ {
		if nums[k] > 0 {
			break
		}
		if k > 0 && nums[k] == nums[k-1] {
			continue
		}
		i, j := k+1, size-1
		for i < j {
			sum := nums[k] + nums[i] + nums[j]
			if sum == 0 {
				ret = append(ret, []int{nums[k], nums[i], nums[j]})
				j--
				for i < j && nums[j] == nums[j+1] {
					j--
				}
				i++
				for i < j && nums[i] == nums[i-1] {
					i++
				}
			} else if sum < 0 {
				i++
				for i < j && nums[i] == nums[i-1] {
					i++
				}
			} else {
				j--
				for i < j && nums[j] == nums[j+1] {
					j--
				}
			}
		}

	}
	return ret
}
// @lc code=end

